File: Chapter4/Main/linalgebra_impl/proc-hom-v.0.2.R


rm(list=ls(all=TRUE))
library(MASS)
library("R.oo")

# Set JAVA_HOME, set max. memory, and load rJava library
Sys.setenv(JAVA_HOME='/usr/java/latest/bin')
options(java.parameters="-Xmx1g")
library(rJava)
.jinit()
print(.jcall("java/lang/System", "S", "getProperty", "java.version"))
library(RJDBC)
drv <- JDBC("org.hsqldb.jdbcDriver","/root/devel/proc_hom/lib/hsqldb.jar", identifier.quote="\"")
conn <- dbConnect(drv, "jdbc:hsqldb:hsql://localhost:9005/proc_hom", "SA", "")


setwd("/root/devel/proc_hom/formulas2")
print(getwd())

#initialize




drop_la_tables <- function(){
  
  
  if(dbExistsTable(conn,"LA_SPECIES")) {
    print("table LA_SPECIES exists -------------------")
    dbRemoveTable(conn,"LA_SPECIES")  
  }
  if(dbExistsTable(conn,"LA_GENES")) {
    print("table LA_SPECIES exists -------------------")
    dbRemoveTable(conn,"LA_GENES")  
  }
  if(dbExistsTable(conn,"LA_ALLELES")) {
    print("table LA_ALLELES exists -------------------")
    dbRemoveTable(conn,"LA_ALLELES")  
  }
  if(dbExistsTable(conn,"LA_ALLELES_SPECIES")) {
    print("table LA_ALLELES_SPECIES exists -------------------")
    dbRemoveTable(conn,"LA_ALLELES_SPECIES")  
  }
  if(dbExistsTable(conn,"LA_PROK_GROUPS")) {
    print("table LA_PROK_GROUPS exists -------------------")
    dbRemoveTable(conn,"LA_PROK_GROUPS")  
  }
  if(dbExistsTable(conn,"LA_SPECIES_PROK_GROUPS")) {
    print("table LA_SPECIES_PROK_GROUPS exists -------------------")
    dbRemoveTable(conn,"LA_SPECIES_PROK_GROUPS")  
  }
  if(dbExistsTable(conn,"LA_ALLELES_GENES")) {
    print("table LA_ALLELES_GENES exists -------------------")
    dbRemoveTable(conn,"LA_ALLELES_GENES")  
  }
  
  if(dbExistsTable(conn,"LA_TRANSFERS")) {
    print("table LA_TRANSFERS exists -------------------")
    dbRemoveTable(conn,"LA_TRANSFERS")  
  }
  
}


create_la_tables <- function() {
  
  #-------------------------------------------------la_species
  sql <- sprintf("create table la_species as (
    select tx.TREE_ORDER - 1 as sp,
    tx.ID as taxon_id,
    tx.TREE_NAME,
    tx.SCI_NAME
    from taxons tx
    order by tx.TREE_ORDER
  ) WITH DATA")
  
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE UNIQUE INDEX la_species_sp ON la_species (sp asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE UNIQUE INDEX la_species_taxon_id ON la_species (taxon_id asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  
  #-------------------------------------genes
  sql <- sprintf("create table la_genes as (
select rownum()-1 gn,
       gn.id as gene_id,
       gn.name
from GENES gn
where gn.name not in ('rbcL')
order by gn.id 
) WITH DATA")
  
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE UNIQUE INDEX la_genes_gn ON la_genes (gn asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE UNIQUE INDEX la_genes_gene_id ON la_genes (gene_id asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE UNIQUE INDEX la_genes_name ON la_genes (name asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  
  
  #-------------------------------------alleles
  
  sql <- sprintf("create table la_alleles as (
select rownum() -1 as al,
       gbs.NCBI_SEQ_ID, 
       ns.VERS_ACCESS
from gene_blo_seqs gbs 
 join NCBI_SEQS ns on ns.ID = gbs.NCBI_SEQ_ID
 --only alleles that have an associated taxon in the tree
 join lA_species ls on ls.TAXON_ID = ns.TAXON_ID
 --only alleles that are in regular genes not rcbL
 join LA_GENES lg on lg.GENE_ID = gbs.GENE_ID
order by gbs.NCBI_SEQ_ID
) WITH DATA")
  
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE UNIQUE INDEX la_alleles_al ON la_alleles (al asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE UNIQUE INDEX la_alleles_ncbi_seq_id ON la_alleles (ncbi_seq_id asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  #-------------------------------------alleles_species
  
  sql <- sprintf("create table la_alleles_species as (
select la.AL,
       ls.SP,
       gbs.NCBI_SEQ_ID, 
       ns.VERS_ACCESS       
from gene_blo_seqs gbs 
 join NCBI_SEQS ns on ns.ID = gbs.NCBI_SEQ_ID
 join LA_ALLELES la on la.NCBI_SEQ_ID = gbs.NCBI_SEQ_ID
 join lA_species ls on ls.TAXON_ID = ns.TAXON_ID
order by la.AL
) WITH DATA")
  
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE UNIQUE INDEX la_alleles_species_al ON la_alleles_species (al asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE INDEX la_alleles_species_sp ON la_alleles_species (sp asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE INDEX la_alleles_species_ncbi_seq_id ON la_alleles_species (ncbi_seq_id asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE INDEX la_alleles_species_vers_access ON la_alleles_species (vers_access asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  
  #-------------------------------------prok-groups
  
  sql <- sprintf("create table la_prok_groups as (
select pg.ORDER_ID as PG,
       pg.id as PROK_GROUP_ID,
       pg.name
from PROK_GROUPS pg
where --pg.ID between 0 and 22
      pg.ID between 23 and 100
order by pg.ORDER_id
) WITH DATA")
  
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE UNIQUE INDEX la_prok_groups_pg ON la_prok_groups (pg asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE UNIQUE INDEX la_prok_groups_ncbi_seq_id ON la_prok_groups (PROK_GROUP_ID asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  #-------------------------------------species-prok-groups
  sql <- sprintf("create table la_species_prok_groups as (
select ls.SP,
       lpg.PG,
       tg.PROK_GROUP_ID,
       tg.WEIGHT_PG,
       tg.TAXON_ID
from TAXON_GROUPS tg
 join LA_PROK_GROUPS lpg on lpg.PROK_GROUP_ID = tg.PROK_GROUP_ID
 join LA_SPECIES ls on ls.TAXON_ID = tg.TAXON_ID
where tg.WEIGHT_PG is not null
      --and tg.PROK_GROUP_ID between 0 and 22
      and tg.PROK_GROUP_ID between 23 and 100
order by ls.SP,
       lpg.PG
) WITH DATA")
  
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE INDEX la_species_prok_groups_sp ON la_species_prok_groups (sp asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE INDEX la_species_prok_groups_pg ON la_species_prok_groups (pg asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE INDEX la_species_prok_groups_prok_group_id ON la_species_prok_groups (prok_group_id asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE INDEX la_species_prok_groups_taxon_id ON la_species_prok_groups (taxon_id asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  
  #-------------------------------------alleles-genes
  sql <- sprintf("create table la_alleles_genes as (
select la.AL,
       lg.GN,
       gbs.gene_id,
       gbs.NCBI_SEQ_ID
from GENE_BLO_SEQS gbs
 join LA_ALLELES la on la.NCBI_SEQ_ID = gbs.NCBI_SEQ_ID
 join LA_GENES lg on lg.GENE_ID = gbs.GENE_ID
order by gbs.GENE_ID
) WITH DATA")
  
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE INDEX la_alleles_genes_al ON la_alleles_genes (al asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE INDEX la_alleles_genes_gn ON la_alleles_genes (gn asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE INDEX la_alleles_genes_gene_id ON la_alleles_genes (gene_id asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  
  sql <- sprintf("CREATE INDEX la_alleles_genes_ncbi_seq_id ON la_alleles_genes (ncbi_seq_id asc)")
  print(sql)
  dbSendUpdate(conn,sql)
  #-------------------------------------transfers
  sql <- sprintf("create table la_transfers as (
select lg.GN,
       la_als.AL as als,
       la_ald.AL as ald,
       hcit.WEIGHT,
       hcif.ID as fragm_id,
       hcif.GENE_ID,
       hcit.SOURCE_ID,
       hcit.DEST_ID       
from HGT_COM_INT_TRANSFERS hcit
 join HGT_COM_INT_FRAGMS hcif on hcif.ID = hcit.HGT_COM_INT_FRAGM_ID
 join LA_ALLELES la_als on la_als.NCBI_SEQ_ID = hcit.SOURCE_ID
 join LA_ALLELES la_ald on la_ald.NCBI_SEQ_ID = hcit.DEST_ID
 join LA_GENES lg on lg.GENE_ID = hcif.GENE_ID
 order by lg.GN,
          hcif.ID 
) WITH DATA")
  
  print(sql)
  dbSendUpdate(conn,sql)
  

}

export_la_tables <- function() {
  
  sql <-sprintf("select * from LA_SPECIES")
  la_species <- dbGetQuery(conn, sql)
  write.matrix(format(la_species, scientific=FALSE), file = "files2/la_species.csv", sep = ",")
  
  sql <-sprintf("select * from LA_GENES")
  la_species <- dbGetQuery(conn, sql)
  write.matrix(format(la_species, scientific=FALSE), file = "files2/la_genes.csv", sep = ",")

  sql <-sprintf("select * from LA_ALLELES")
  la_species <- dbGetQuery(conn, sql)
  write.matrix(format(la_species, scientific=FALSE), file = "files2/la_alleles.csv", sep = ",")
  
  sql <-sprintf("select * from LA_ALLELES_SPECIES")
  la_species <- dbGetQuery(conn, sql)
  write.matrix(format(la_species, scientific=FALSE), file = "files2/la_alleles_species.csv", sep = ",")
  
  sql <-sprintf("select * from LA_PROK_GROUPS")
  la_species <- dbGetQuery(conn, sql)
  write.matrix(format(la_species, scientific=FALSE), file = "files2/la_prok_groups.csv", sep = ",")
  
  sql <-sprintf("select * from LA_SPECIES_PROK_GROUPS")
  la_species <- dbGetQuery(conn, sql)
  write.matrix(format(la_species, scientific=FALSE), file = "files2/la_species_prok_groups.csv", sep = ",")
  
  sql <-sprintf("select * from LA_ALLELES_GENES")
  la_species <- dbGetQuery(conn, sql)
  write.matrix(format(la_species, scientific=FALSE), file = "files2/la_alleles_genes.csv", sep = ",")
  
  sql <-sprintf("select * from LA_TRANSFERS")
  la_species <- dbGetQuery(conn, sql)
  write.matrix(format(la_species, scientific=FALSE), file = "files2/la_transfers.csv", sep = ",")
  
  
}

result <- drop_la_tables()
print(paste("drop_la_tables",result))

result <- create_la_tables()
print(paste("create_la_tables",result))

result <- export_la_tables()
print(paste("export_la_tables",result))

File: Chapter4/Main/linalgebra_impl/proc-hom-v.0.2.py
#!/usr/bin/env python

# -*- coding: utf-8 -*-

"""
Created on Sun Sep 21 12:37:33 2014

@author: Dunarel Badescu
"""

import csv
import numpy as np

    
from numpy import array, dot
from scipy import sparse

#

#PG,PROK_GROUP_ID,NAME
la_prok_groups= np.genfromtxt("files2/la_prok_groups.csv", dtype="i8,i8,|S35",  delimiter=',', names=True) 
prok_groups =  la_prok_groups['PG']
print "prok_groups.size: ",prok_groups.size

#SP,TAXON_ID,TREE_NAME,SCI_NAME
la_species= np.genfromtxt("files2/la_species.csv", dtype="i8,i8,|S35,|S35",  delimiter=',', names=True) 
species =  la_species['SP']
print "species.size: ",species.size


#AL,NCBI_SEQ_ID,VERS_ACCESS
la_alleles= np.genfromtxt("files2/la_alleles.csv", dtype="i8,i8,|S35",  delimiter=',', names=True) 

alleles =  la_alleles['AL']

print "alleles.size: ",alleles.size

#print data['NCBI_SEQ_ID']
#print data['VERS_ACCESS'][0]


#GN,GENE_ID,NAME
la_genes= np.genfromtxt("files2/la_genes.csv", dtype="i8,i8,|S35",  delimiter=',', names=True) 
genes =  la_genes['GN']

print "genes.size: ",genes.size

#AL,GN,GENE_ID,NCBI_SEQ_ID
la_alleles_genes = np.genfromtxt("files2/la_alleles_genes.csv", dtype="i8,i8,i8,i8",  delimiter=',', names=True) 

L = sparse.csr_matrix((alleles.size,genes.size), dtype="f8")

for row in la_alleles_genes:
    al = row[0]
    gn = row[1]
    #print "sp, al: ",sp,al
    L[al,gn]=1.0

#AL,SP,NCBI_SEQ_ID,VERS_ACCESS
la_alleles_species = np.genfromtxt("files2/la_alleles_species.csv", dtype="i8,i8,i8,|S35",  delimiter=',', names=True) 

Z = sparse.csr_matrix((species.size,alleles.size), dtype="f8")
print Z[1,1]

for row in la_alleles_species:
    sp = row[1]
    al = row[0]
    #print "sp, al: ",sp,al
    Z[sp,al]=1.0

#print "Z: ",Z

#SP,PG,PROK_GROUP_ID,WEIGHT_PG,TAXON_ID
la_species_prok_groups = np.genfromtxt("files2/la_species_prok_groups.csv", dtype="i8,i8,i8,f8,i8",  delimiter=',', names=True) 

W = sparse.csr_matrix((prok_groups.size,species.size), dtype="f8")
#print Z[1,1]

for row in la_species_prok_groups:
    sp = row[0]    
    pg = row[1]
    wt = row[3]
    #print "pg, sp, wt: ",pg,sp,wt
    W[pg,sp]=wt
    
    
#print "W: ",W

#PxG
PG = W.dot(Z).dot(L)

for i in range(0,prok_groups.size):
    pgsn = PG.sum(axis=1)[i]
    #print "i,pgsn: ",i,pgsn
    
#GN,ALS,ALD,WEIGHT,FRAGM_ID,GENE_ID,SOURCE_ID,DEST_ID
la_transfers = np.genfromtxt("files2/la_transfers.csv", dtype="i8,i8,i8,f8,i8,i8,i8,i8",  delimiter=',', names=True) 

#R dense numpy 3d matrix gene x proc x proc
R = np.zeros((genes.size,prok_groups.size,prok_groups.size), dtype="f8")

# dictionnary of sparse matrices
KK=[]
for i in range(0, genes.size):
    #print "We're on time %d" % (i)
    KK.append(sparse.csr_matrix((alleles.size,alleles.size), dtype="f8"))
    
#process transfers into KK
for row in la_transfers:
    gn = row[0]
    als = row[1]
    ald = row[2]
    wt = row[3]
    #print "sp, al: ",sp,al
    KK[gn][als,ald]+=wt

#collect results in R
for i in range(0, genes.size):
    #R.append(W.dot(Z).dot(KK[i]).dot(Z.T).dot(W.T))
    R[i] = W.dot(Z).dot(KK[i]).dot(Z.T).dot(W.T).todense()
    
    #rshape = R.shape
    #print rshape[0],rshape[1]
#print "R: ",R,R.__len__()

def k_div_nm(psrc, pdst):
    
    cor_perc = 100.0
    
    absol = 0.0
    rel = 0.0
    nb_genes = 0.0

    abs_sum = 0.0
    rel_sum = 0.0


    for gn in range(0, genes.size):
        #R.append(W.dot(Z).dot(KK[i]).dot(Z.T).dot(W.T))
        #init case values
        #val_elem = 0.0
        #value for gene
        vfg = R[gn][psrc, pdst]
        #size of gene source
        sgs = PG[psrc,gn]
        #size of gene dest
        sgd = PG[pdst,gn]
        den = (sgs*(sgd-1))/2 if (psrc == pdst) else (sgs * sgd)
  
  
        if den != 0:
            nb_genes +=1 
            #local variables
            abs_elem =  vfg
            rel_elem = (vfg / den)
            #cummulative
            abs_sum += abs_elem 
            rel_sum += rel_elem

    #normalization
    absol = abs_sum
    if nb_genes != 0:
        rel = (rel_sum * cor_perc ) /  nb_genes 
    
    
    return [absol, rel, nb_genes]


#
abso = np.zeros((prok_groups.size,prok_groups.size), dtype="f8")
rel =  np.zeros((prok_groups.size,prok_groups.size), dtype="f8")
ngenes =  np.zeros((prok_groups.size,prok_groups.size), dtype="f8")
#
for psrc in range (0,prok_groups.size):
    for pdst in range (0,prok_groups.size):
        abso[psrc,pdst], rel[psrc,pdst], ngenes[psrc,pdst] = k_div_nm(psrc, pdst)
        #pass

#print "abso",abso
#np.savetxt('exports/abso.csv', abso, delimiter=',', fmt='%20.10f')   # use exponential notation
#np.savetxt('exports/rel.csv', rel, delimiter=',', fmt='%20.10f')   # use exponential notation
#np.savetxt('exports/ngenes.csv', ngenes, delimiter=',', fmt='%20.10f')   # use exponential notation        
        
tt_xsums = np.zeros((prok_groups.size), dtype="f8")
 
def calc_transf_gr_total_one_dim():
    tt_xsums = []
    #row by row
    for s in range(0,prok_groups.size):
         
      row_abs = abso[s,]
      row_gn  = ngenes[s,]
      #print "row_abs: ",row_abs
      
      row_sum = 0.0
      #for d in range(0,prok_groups.size):
      #    row_sum += row_abs[d]
      
      pgsn = PG.sum(axis=1)[s]
      print np.squeeze(pgsn).shape
      row_tt = row_sum / pgsn * 100 / 1.1 #correct to 110 genes
      print np.squeeze(row_tt).shape
      print "row_tt: ", row_tt,row_tt.shape
    
#calc_transf_gr_total_one_dim()
PGSN = PG.sum(axis=1).T
#WTS = abso.sum(axis=0) / PGSN * 100 / 1.1
#WTD = abso.sum(axis=1) / PGSN * 100 / 1.1

WTS = abso.sum(axis=0) / PGSN * 100 
WTD = abso.sum(axis=1) / PGSN * 100 


PGSN = PGSN.A1
WTS = WTS.A1
WTD = WTD.A1


MEAN_HGT_SRC = (WTS*PGSN).sum()/PGSN.sum()
MEAN_HGT_DST = (WTD*PGSN).sum()/PGSN.sum()

print "MEAN_HGT_SRC: ", MEAN_HGT_SRC
print "MEAN_HGT_DST: ", MEAN_HGT_DST


